// @cinelink/networking/gen/client/dist/mjs/example.client.js
// @generated by rootSdk Generator
// DO NOT HAND EDIT
// tslint: disable
import { AdvanceQueueRequest, DecideHostClaimRequest, FetchSubtitleRequest, FetchSubtitleResponse, GetRoomStateRequest, GetRoomStateResponse, LeaveRoomRequest, ListRoomsRequest, ListRoomsResponse, MoveQueueItemRequest, MutationResponse, PlaybackRequest, PlaylistMutationRequest, RemoveQueueItemRequest, RequestHostClaimRequest, SearchSubtitlesRequest, SearchSubtitlesResponse, SetDurationRequest, SetHostRequest, SetMediaRequest, SetRateRequest, SetRoomNameRequest, SetSubtitleRequest, SetViewerQueuePolicyRequest, StateChangedEvent } from "@cinelink/gen-shared";
import { EventEmitter, rootClient } from "@rootsdk/client-app";
export var CineLinkServiceClientEvent;
(function (CineLinkServiceClientEvent) {
    CineLinkServiceClientEvent["StateChanged"] = "broadcastStateChanged";
})(CineLinkServiceClientEvent || (CineLinkServiceClientEvent = {}));
export class CineLinkServiceClient extends EventEmitter {
    constructor() { super(); }
    get ClientServiceName() {
        return "CineLinkServiceClient";
    }
    listRooms(request) {
        return rootClient.sendWithResponse('/.CineLinkService/ListRooms', request, this.listRooms);
    }
    getRoomState(request) {
        return rootClient.sendWithResponse('/.CineLinkService/GetRoomState', request, this.getRoomState);
    }
    leaveRoom(request) {
        return rootClient.sendWithResponse('/.CineLinkService/LeaveRoom', request, this.leaveRoom);
    }
    setHost(request) {
        return rootClient.sendWithResponse('/.CineLinkService/SetHost', request, this.setHost);
    }
    setRoomName(request) {
        return rootClient.sendWithResponse('/.CineLinkService/SetRoomName', request, this.setRoomName);
    }
    setViewerQueuePolicy(request) {
        return rootClient.sendWithResponse('/.CineLinkService/SetViewerQueuePolicy', request, this.setViewerQueuePolicy);
    }
    requestHostClaim(request) {
        return rootClient.sendWithResponse('/.CineLinkService/RequestHostClaim', request, this.requestHostClaim);
    }
    decideHostClaim(request) {
        return rootClient.sendWithResponse('/.CineLinkService/DecideHostClaim', request, this.decideHostClaim);
    }
    setMedia(request) {
        return rootClient.sendWithResponse('/.CineLinkService/SetMedia', request, this.setMedia);
    }
    addQueueNext(request) {
        return rootClient.sendWithResponse('/.CineLinkService/AddQueueNext', request, this.addQueueNext);
    }
    addQueueLast(request) {
        return rootClient.sendWithResponse('/.CineLinkService/AddQueueLast', request, this.addQueueLast);
    }
    advanceQueue(request) {
        return rootClient.sendWithResponse('/.CineLinkService/AdvanceQueue', request, this.advanceQueue);
    }
    previousQueue(request) {
        return rootClient.sendWithResponse('/.CineLinkService/PreviousQueue', request, this.previousQueue);
    }
    removeQueueItem(request) {
        return rootClient.sendWithResponse('/.CineLinkService/RemoveQueueItem', request, this.removeQueueItem);
    }
    moveQueueItem(request) {
        return rootClient.sendWithResponse('/.CineLinkService/MoveQueueItem', request, this.moveQueueItem);
    }
    setDuration(request) {
        return rootClient.sendWithResponse('/.CineLinkService/SetDuration', request, this.setDuration);
    }
    play(request) {
        return rootClient.sendWithResponse('/.CineLinkService/Play', request, this.play);
    }
    pause(request) {
        return rootClient.sendWithResponse('/.CineLinkService/Pause', request, this.pause);
    }
    seek(request) {
        return rootClient.sendWithResponse('/.CineLinkService/Seek', request, this.seek);
    }
    setRate(request) {
        return rootClient.sendWithResponse('/.CineLinkService/SetRate', request, this.setRate);
    }
    searchSubtitles(request) {
        return rootClient.sendWithResponse('/.CineLinkService/SearchSubtitles', request, this.searchSubtitles);
    }
    fetchSubtitle(request) {
        return rootClient.sendWithResponse('/.CineLinkService/FetchSubtitle', request, this.fetchSubtitle);
    }
    setSubtitle(request) {
        return rootClient.sendWithResponse('/.CineLinkService/SetSubtitle', request, this.setSubtitle);
    }
    __register() {
        return [{
                serviceName: 'CineLinkServiceClient',
                methodName: 'ListRooms',
                path: '/.CineLinkService/ListRooms',
                requestSerialize: value => ListRoomsRequest.toBinary(value),
                responseDeserialize: bytes => ListRoomsResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'GetRoomState',
                path: '/.CineLinkService/GetRoomState',
                requestSerialize: value => GetRoomStateRequest.toBinary(value),
                responseDeserialize: bytes => GetRoomStateResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'LeaveRoom',
                path: '/.CineLinkService/LeaveRoom',
                requestSerialize: value => LeaveRoomRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'SetHost',
                path: '/.CineLinkService/SetHost',
                requestSerialize: value => SetHostRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'SetRoomName',
                path: '/.CineLinkService/SetRoomName',
                requestSerialize: value => SetRoomNameRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'SetViewerQueuePolicy',
                path: '/.CineLinkService/SetViewerQueuePolicy',
                requestSerialize: value => SetViewerQueuePolicyRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'RequestHostClaim',
                path: '/.CineLinkService/RequestHostClaim',
                requestSerialize: value => RequestHostClaimRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'DecideHostClaim',
                path: '/.CineLinkService/DecideHostClaim',
                requestSerialize: value => DecideHostClaimRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'SetMedia',
                path: '/.CineLinkService/SetMedia',
                requestSerialize: value => SetMediaRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'AddQueueNext',
                path: '/.CineLinkService/AddQueueNext',
                requestSerialize: value => PlaylistMutationRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'AddQueueLast',
                path: '/.CineLinkService/AddQueueLast',
                requestSerialize: value => PlaylistMutationRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'AdvanceQueue',
                path: '/.CineLinkService/AdvanceQueue',
                requestSerialize: value => AdvanceQueueRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'PreviousQueue',
                path: '/.CineLinkService/PreviousQueue',
                requestSerialize: value => AdvanceQueueRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'RemoveQueueItem',
                path: '/.CineLinkService/RemoveQueueItem',
                requestSerialize: value => RemoveQueueItemRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'MoveQueueItem',
                path: '/.CineLinkService/MoveQueueItem',
                requestSerialize: value => MoveQueueItemRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'SetDuration',
                path: '/.CineLinkService/SetDuration',
                requestSerialize: value => SetDurationRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'Play',
                path: '/.CineLinkService/Play',
                requestSerialize: value => PlaybackRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'Pause',
                path: '/.CineLinkService/Pause',
                requestSerialize: value => PlaybackRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'Seek',
                path: '/.CineLinkService/Seek',
                requestSerialize: value => PlaybackRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'SetRate',
                path: '/.CineLinkService/SetRate',
                requestSerialize: value => SetRateRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'SearchSubtitles',
                path: '/.CineLinkService/SearchSubtitles',
                requestSerialize: value => SearchSubtitlesRequest.toBinary(value),
                responseDeserialize: bytes => SearchSubtitlesResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'FetchSubtitle',
                path: '/.CineLinkService/FetchSubtitle',
                requestSerialize: value => FetchSubtitleRequest.toBinary(value),
                responseDeserialize: bytes => FetchSubtitleResponse.fromBinary(bytes)
            },
            {
                serviceName: 'CineLinkServiceClient',
                methodName: 'SetSubtitle',
                path: '/.CineLinkService/SetSubtitle',
                requestSerialize: value => SetSubtitleRequest.toBinary(value),
                responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
            },
            {
                serviceName: "CineLinkServiceClient",
                methodName: "BroadcastStateChanged",
                path: '/.CineLinkService/BroadcastStateChanged',
                funct: (e) => { this.emit(CineLinkServiceClientEvent.StateChanged, e); },
                responseDeserialize: bytes => StateChangedEvent.fromBinary(bytes)
            },
        ];
    }
}
export const cineLinkServiceClient = new CineLinkServiceClient();
rootClient.addClient(cineLinkServiceClient);
