// @cinelink/networking/gen/client/src/example.client.ts
// @generated by rootSdk Generator
// DO NOT HAND EDIT
// tslint: disable

import {
  AdvanceQueueRequest,
	DecideHostClaimRequest,
	FetchSubtitleRequest,
	FetchSubtitleResponse,
	GetRoomStateRequest,
	GetRoomStateResponse,
	LeaveRoomRequest,
	ListRoomsRequest,
	ListRoomsResponse,
	MoveQueueItemRequest,
	MutationResponse,
	PlaybackRequest,
	PlaylistMutationRequest,
	RemoveQueueItemRequest,
	RequestHostClaimRequest,
	SearchSubtitlesRequest,
	SearchSubtitlesResponse,
	SetDurationRequest,
	SetHostRequest,
	SetMediaRequest,
	SetRateRequest,
	SetRoomNameRequest,
	SetSubtitleRequest,
	SetViewerQueuePolicyRequest,
	StateChangedEvent,
	Void
} from "@cinelink/gen-shared"
import {
  EventEmitter,
	IRootClient,
	RootClientService,
	TypedEventEmitter,
	UntypedClientMethodDefinition,
	rootClient
} from "@rootsdk/client-app"

export enum CineLinkServiceClientEvent {
		StateChanged = 'broadcastStateChanged'
}
export type CineLinkServiceClientEvents = {
		'broadcastStateChanged': (event: StateChangedEvent) => void
}

export class CineLinkServiceClient extends (EventEmitter as new() => TypedEventEmitter<CineLinkServiceClientEvents>) implements RootClientService
 {
  constructor() { super(); }
 
	public get ClientServiceName(): string {
		return "CineLinkServiceClient";
	}

  listRooms(request: ListRoomsRequest):Promise<ListRoomsResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/ListRooms',request, this.listRooms);
	}
	getRoomState(request: GetRoomStateRequest):Promise<GetRoomStateResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/GetRoomState',request, this.getRoomState);
	}
	leaveRoom(request: LeaveRoomRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/LeaveRoom',request, this.leaveRoom);
	}
	setHost(request: SetHostRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/SetHost',request, this.setHost);
	}
	setRoomName(request: SetRoomNameRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/SetRoomName',request, this.setRoomName);
	}
	setViewerQueuePolicy(request: SetViewerQueuePolicyRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/SetViewerQueuePolicy',request, this.setViewerQueuePolicy);
	}
	requestHostClaim(request: RequestHostClaimRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/RequestHostClaim',request, this.requestHostClaim);
	}
	decideHostClaim(request: DecideHostClaimRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/DecideHostClaim',request, this.decideHostClaim);
	}
	setMedia(request: SetMediaRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/SetMedia',request, this.setMedia);
	}
	addQueueNext(request: PlaylistMutationRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/AddQueueNext',request, this.addQueueNext);
	}
	addQueueLast(request: PlaylistMutationRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/AddQueueLast',request, this.addQueueLast);
	}
	advanceQueue(request: AdvanceQueueRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/AdvanceQueue',request, this.advanceQueue);
	}
	previousQueue(request: AdvanceQueueRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/PreviousQueue',request, this.previousQueue);
	}
	removeQueueItem(request: RemoveQueueItemRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/RemoveQueueItem',request, this.removeQueueItem);
	}
	moveQueueItem(request: MoveQueueItemRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/MoveQueueItem',request, this.moveQueueItem);
	}
	setDuration(request: SetDurationRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/SetDuration',request, this.setDuration);
	}
	play(request: PlaybackRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/Play',request, this.play);
	}
	pause(request: PlaybackRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/Pause',request, this.pause);
	}
	seek(request: PlaybackRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/Seek',request, this.seek);
	}
	setRate(request: SetRateRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/SetRate',request, this.setRate);
	}
	searchSubtitles(request: SearchSubtitlesRequest):Promise<SearchSubtitlesResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/SearchSubtitles',request, this.searchSubtitles);
	}
	fetchSubtitle(request: FetchSubtitleRequest):Promise<FetchSubtitleResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/FetchSubtitle',request, this.fetchSubtitle);
	}
	setSubtitle(request: SetSubtitleRequest):Promise<MutationResponse> {
		return (<any>rootClient).sendWithResponse('/.CineLinkService/SetSubtitle',request, this.setSubtitle);
	}
	private __register(): UntypedClientMethodDefinition[] {
		return [	{
      serviceName: 'CineLinkServiceClient',
			methodName: 'ListRooms',
			path: '/.CineLinkService/ListRooms',
			requestSerialize: value => ListRoomsRequest.toBinary(value),
			responseDeserialize: bytes => ListRoomsResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'GetRoomState',
			path: '/.CineLinkService/GetRoomState',
			requestSerialize: value => GetRoomStateRequest.toBinary(value),
			responseDeserialize: bytes => GetRoomStateResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'LeaveRoom',
			path: '/.CineLinkService/LeaveRoom',
			requestSerialize: value => LeaveRoomRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'SetHost',
			path: '/.CineLinkService/SetHost',
			requestSerialize: value => SetHostRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'SetRoomName',
			path: '/.CineLinkService/SetRoomName',
			requestSerialize: value => SetRoomNameRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'SetViewerQueuePolicy',
			path: '/.CineLinkService/SetViewerQueuePolicy',
			requestSerialize: value => SetViewerQueuePolicyRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'RequestHostClaim',
			path: '/.CineLinkService/RequestHostClaim',
			requestSerialize: value => RequestHostClaimRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'DecideHostClaim',
			path: '/.CineLinkService/DecideHostClaim',
			requestSerialize: value => DecideHostClaimRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'SetMedia',
			path: '/.CineLinkService/SetMedia',
			requestSerialize: value => SetMediaRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'AddQueueNext',
			path: '/.CineLinkService/AddQueueNext',
			requestSerialize: value => PlaylistMutationRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'AddQueueLast',
			path: '/.CineLinkService/AddQueueLast',
			requestSerialize: value => PlaylistMutationRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'AdvanceQueue',
			path: '/.CineLinkService/AdvanceQueue',
			requestSerialize: value => AdvanceQueueRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'PreviousQueue',
			path: '/.CineLinkService/PreviousQueue',
			requestSerialize: value => AdvanceQueueRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'RemoveQueueItem',
			path: '/.CineLinkService/RemoveQueueItem',
			requestSerialize: value => RemoveQueueItemRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'MoveQueueItem',
			path: '/.CineLinkService/MoveQueueItem',
			requestSerialize: value => MoveQueueItemRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'SetDuration',
			path: '/.CineLinkService/SetDuration',
			requestSerialize: value => SetDurationRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'Play',
			path: '/.CineLinkService/Play',
			requestSerialize: value => PlaybackRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'Pause',
			path: '/.CineLinkService/Pause',
			requestSerialize: value => PlaybackRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'Seek',
			path: '/.CineLinkService/Seek',
			requestSerialize: value => PlaybackRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'SetRate',
			path: '/.CineLinkService/SetRate',
			requestSerialize: value => SetRateRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'SearchSubtitles',
			path: '/.CineLinkService/SearchSubtitles',
			requestSerialize: value => SearchSubtitlesRequest.toBinary(value),
			responseDeserialize: bytes => SearchSubtitlesResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'FetchSubtitle',
			path: '/.CineLinkService/FetchSubtitle',
			requestSerialize: value => FetchSubtitleRequest.toBinary(value),
			responseDeserialize: bytes => FetchSubtitleResponse.fromBinary(bytes)
    },
		{
      serviceName: 'CineLinkServiceClient',
			methodName: 'SetSubtitle',
			path: '/.CineLinkService/SetSubtitle',
			requestSerialize: value => SetSubtitleRequest.toBinary(value),
			responseDeserialize: bytes => MutationResponse.fromBinary(bytes)
    },
		{
			serviceName: "CineLinkServiceClient",
      methodName: "BroadcastStateChanged",
      path: '/.CineLinkService/BroadcastStateChanged',
      funct: (e: StateChangedEvent) => { this.emit(CineLinkServiceClientEvent.StateChanged,e) },
      responseDeserialize: bytes => StateChangedEvent.fromBinary(bytes)
    },
	]}
}

export const cineLinkServiceClient = new CineLinkServiceClient();

(<IRootClient><unknown>rootClient).addClient(cineLinkServiceClient);
