// @cinelink/gen-server/src/example.server.ts
// @generated by rootSdk Generator
// DO NOT HAND EDIT
// tslint: disable

import {
  AdvanceQueueRequest,
	DecideHostClaimRequest,
	FetchSubtitleRequest,
	FetchSubtitleResponse,
	GetRoomStateRequest,
	GetRoomStateResponse,
	HostLaunchSyncRequest,
	HostStartSyncRequest,
	LeaveRoomRequest,
	ListRoomsRequest,
	ListRoomsResponse,
	MoveQueueItemRequest,
	MutationResponse,
	PlaybackRequest,
	PlaylistMutationRequest,
	RemoveQueueItemRequest,
	ReportSyncStatusRequest,
	RequestHostClaimRequest,
	SearchSubtitlesRequest,
	SearchSubtitlesResponse,
	SetDurationRequest,
	SetHostRequest,
	SetMediaRequest,
	SetRateRequest,
	SetRoomNameRequest,
	SetSubtitleRequest,
	SetViewerQueuePolicyRequest,
	StateChangedEvent,
	Void
} from "@cinelink/gen-shared"
import {
  BroadcastClientMerged,
	Client,
	DeviceContext,
	RootServer,
	RootServerService,
	UntypedServerMethodDefinition,
	rootServer
} from "@rootsdk/server-app"

export abstract class CineLinkServiceBase implements RootServerService {
	abstract listRooms(request: ListRoomsRequest, client: Client): Promise<ListRoomsResponse>;
	abstract getRoomState(request: GetRoomStateRequest, client: Client): Promise<GetRoomStateResponse>;
	abstract leaveRoom(request: LeaveRoomRequest, client: Client): Promise<MutationResponse>;
	abstract setHost(request: SetHostRequest, client: Client): Promise<MutationResponse>;
	abstract setRoomName(request: SetRoomNameRequest, client: Client): Promise<MutationResponse>;
	abstract setViewerQueuePolicy(request: SetViewerQueuePolicyRequest, client: Client): Promise<MutationResponse>;
	abstract requestHostClaim(request: RequestHostClaimRequest, client: Client): Promise<MutationResponse>;
	abstract decideHostClaim(request: DecideHostClaimRequest, client: Client): Promise<MutationResponse>;
	abstract setMedia(request: SetMediaRequest, client: Client): Promise<MutationResponse>;
	abstract addQueueNext(request: PlaylistMutationRequest, client: Client): Promise<MutationResponse>;
	abstract addQueueLast(request: PlaylistMutationRequest, client: Client): Promise<MutationResponse>;
	abstract advanceQueue(request: AdvanceQueueRequest, client: Client): Promise<MutationResponse>;
	abstract previousQueue(request: AdvanceQueueRequest, client: Client): Promise<MutationResponse>;
	abstract removeQueueItem(request: RemoveQueueItemRequest, client: Client): Promise<MutationResponse>;
	abstract moveQueueItem(request: MoveQueueItemRequest, client: Client): Promise<MutationResponse>;
	abstract setDuration(request: SetDurationRequest, client: Client): Promise<MutationResponse>;
	abstract play(request: PlaybackRequest, client: Client): Promise<MutationResponse>;
	abstract pause(request: PlaybackRequest, client: Client): Promise<MutationResponse>;
	abstract seek(request: PlaybackRequest, client: Client): Promise<MutationResponse>;
	abstract setRate(request: SetRateRequest, client: Client): Promise<MutationResponse>;
	abstract searchSubtitles(request: SearchSubtitlesRequest, client: Client): Promise<SearchSubtitlesResponse>;
	abstract fetchSubtitle(request: FetchSubtitleRequest, client: Client): Promise<FetchSubtitleResponse>;
	abstract setSubtitle(request: SetSubtitleRequest, client: Client): Promise<MutationResponse>;
	abstract reportSyncStatus(request: ReportSyncStatusRequest, client: Client): Promise<MutationResponse>;
	abstract hostStartSync(request: HostStartSyncRequest, client: Client): Promise<MutationResponse>;
	abstract hostLaunchSync(request: HostLaunchSyncRequest, client: Client): Promise<MutationResponse>;
	public broadcastStateChanged(message: StateChangedEvent, clients: BroadcastClientMerged, except?: DeviceContext) {
		(<any>rootServer).sendBroadcast('/.CineLinkService/BroadcastStateChanged', message, clients, except);
	}

	private __register(): UntypedServerMethodDefinition[] {
		return [	{
			serviceName: 'CineLinkService',
			methodName: 'ListRooms',
			funct: (e: ListRoomsRequest, c: Client) => { return this.listRooms(e, c) },
			path: '/.CineLinkService/ListRooms',
			requestDeserialize: bytes => ListRoomsRequest.fromBinary(bytes),
			responseSerialize: value => ListRoomsResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'GetRoomState',
			funct: (e: GetRoomStateRequest, c: Client) => { return this.getRoomState(e, c) },
			path: '/.CineLinkService/GetRoomState',
			requestDeserialize: bytes => GetRoomStateRequest.fromBinary(bytes),
			responseSerialize: value => GetRoomStateResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'LeaveRoom',
			funct: (e: LeaveRoomRequest, c: Client) => { return this.leaveRoom(e, c) },
			path: '/.CineLinkService/LeaveRoom',
			requestDeserialize: bytes => LeaveRoomRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'SetHost',
			funct: (e: SetHostRequest, c: Client) => { return this.setHost(e, c) },
			path: '/.CineLinkService/SetHost',
			requestDeserialize: bytes => SetHostRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'SetRoomName',
			funct: (e: SetRoomNameRequest, c: Client) => { return this.setRoomName(e, c) },
			path: '/.CineLinkService/SetRoomName',
			requestDeserialize: bytes => SetRoomNameRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'SetViewerQueuePolicy',
			funct: (e: SetViewerQueuePolicyRequest, c: Client) => { return this.setViewerQueuePolicy(e, c) },
			path: '/.CineLinkService/SetViewerQueuePolicy',
			requestDeserialize: bytes => SetViewerQueuePolicyRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'RequestHostClaim',
			funct: (e: RequestHostClaimRequest, c: Client) => { return this.requestHostClaim(e, c) },
			path: '/.CineLinkService/RequestHostClaim',
			requestDeserialize: bytes => RequestHostClaimRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'DecideHostClaim',
			funct: (e: DecideHostClaimRequest, c: Client) => { return this.decideHostClaim(e, c) },
			path: '/.CineLinkService/DecideHostClaim',
			requestDeserialize: bytes => DecideHostClaimRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'SetMedia',
			funct: (e: SetMediaRequest, c: Client) => { return this.setMedia(e, c) },
			path: '/.CineLinkService/SetMedia',
			requestDeserialize: bytes => SetMediaRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'AddQueueNext',
			funct: (e: PlaylistMutationRequest, c: Client) => { return this.addQueueNext(e, c) },
			path: '/.CineLinkService/AddQueueNext',
			requestDeserialize: bytes => PlaylistMutationRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'AddQueueLast',
			funct: (e: PlaylistMutationRequest, c: Client) => { return this.addQueueLast(e, c) },
			path: '/.CineLinkService/AddQueueLast',
			requestDeserialize: bytes => PlaylistMutationRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'AdvanceQueue',
			funct: (e: AdvanceQueueRequest, c: Client) => { return this.advanceQueue(e, c) },
			path: '/.CineLinkService/AdvanceQueue',
			requestDeserialize: bytes => AdvanceQueueRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'PreviousQueue',
			funct: (e: AdvanceQueueRequest, c: Client) => { return this.previousQueue(e, c) },
			path: '/.CineLinkService/PreviousQueue',
			requestDeserialize: bytes => AdvanceQueueRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'RemoveQueueItem',
			funct: (e: RemoveQueueItemRequest, c: Client) => { return this.removeQueueItem(e, c) },
			path: '/.CineLinkService/RemoveQueueItem',
			requestDeserialize: bytes => RemoveQueueItemRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'MoveQueueItem',
			funct: (e: MoveQueueItemRequest, c: Client) => { return this.moveQueueItem(e, c) },
			path: '/.CineLinkService/MoveQueueItem',
			requestDeserialize: bytes => MoveQueueItemRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'SetDuration',
			funct: (e: SetDurationRequest, c: Client) => { return this.setDuration(e, c) },
			path: '/.CineLinkService/SetDuration',
			requestDeserialize: bytes => SetDurationRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'Play',
			funct: (e: PlaybackRequest, c: Client) => { return this.play(e, c) },
			path: '/.CineLinkService/Play',
			requestDeserialize: bytes => PlaybackRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'Pause',
			funct: (e: PlaybackRequest, c: Client) => { return this.pause(e, c) },
			path: '/.CineLinkService/Pause',
			requestDeserialize: bytes => PlaybackRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'Seek',
			funct: (e: PlaybackRequest, c: Client) => { return this.seek(e, c) },
			path: '/.CineLinkService/Seek',
			requestDeserialize: bytes => PlaybackRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'SetRate',
			funct: (e: SetRateRequest, c: Client) => { return this.setRate(e, c) },
			path: '/.CineLinkService/SetRate',
			requestDeserialize: bytes => SetRateRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'SearchSubtitles',
			funct: (e: SearchSubtitlesRequest, c: Client) => { return this.searchSubtitles(e, c) },
			path: '/.CineLinkService/SearchSubtitles',
			requestDeserialize: bytes => SearchSubtitlesRequest.fromBinary(bytes),
			responseSerialize: value => SearchSubtitlesResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'FetchSubtitle',
			funct: (e: FetchSubtitleRequest, c: Client) => { return this.fetchSubtitle(e, c) },
			path: '/.CineLinkService/FetchSubtitle',
			requestDeserialize: bytes => FetchSubtitleRequest.fromBinary(bytes),
			responseSerialize: value => FetchSubtitleResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'SetSubtitle',
			funct: (e: SetSubtitleRequest, c: Client) => { return this.setSubtitle(e, c) },
			path: '/.CineLinkService/SetSubtitle',
			requestDeserialize: bytes => SetSubtitleRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'ReportSyncStatus',
			funct: (e: ReportSyncStatusRequest, c: Client) => { return this.reportSyncStatus(e, c) },
			path: '/.CineLinkService/ReportSyncStatus',
			requestDeserialize: bytes => ReportSyncStatusRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'HostStartSync',
			funct: (e: HostStartSyncRequest, c: Client) => { return this.hostStartSync(e, c) },
			path: '/.CineLinkService/HostStartSync',
			requestDeserialize: bytes => HostStartSyncRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: 'CineLinkService',
			methodName: 'HostLaunchSync',
			funct: (e: HostLaunchSyncRequest, c: Client) => { return this.hostLaunchSync(e, c) },
			path: '/.CineLinkService/HostLaunchSync',
			requestDeserialize: bytes => HostLaunchSyncRequest.fromBinary(bytes),
			responseSerialize: value => MutationResponse.toBinary(value)
		},
		{
			serviceName: "CineLinkService",
			methodName: "BroadcastStateChanged",
      path: '/.CineLinkService/BroadcastStateChanged',
      requestSerialize: value => StateChangedEvent.toBinary(value)
    },
	]}

	public get ServerServiceName(): string {
		return "CineLinkService";
	}

}
